File: app/src/main/java/org/ole/planet/myplanet/ui/course/AdapterCourses.java
package org.ole.planet.myplanet.ui.course;

import android.content.Context;
import android.os.Build;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.MotionEvent;
import android.view.View;
import android.view.ViewGroup;
import android.widget.CheckBox;
import android.widget.SeekBar;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.appcompat.widget.AppCompatRatingBar;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.RecyclerView;

import com.google.android.flexbox.FlexboxLayout;
import com.google.gson.JsonObject;

import org.ole.planet.myplanet.R;
import org.ole.planet.myplanet.callback.OnCourseItemSelected;
import org.ole.planet.myplanet.callback.OnHomeItemClickListener;
import org.ole.planet.myplanet.callback.OnRatingChangeListener;
import org.ole.planet.myplanet.databinding.RowCourseBinding;
import org.ole.planet.myplanet.model.RealmMyCourse;
import org.ole.planet.myplanet.model.RealmTag;
import org.ole.planet.myplanet.utilities.JsonUtils;
import org.ole.planet.myplanet.utilities.TimeUtils;
import org.ole.planet.myplanet.utilities.Utilities;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.HashMap;
import java.util.List;

import fisk.chipcloud.ChipCloud;
import fisk.chipcloud.ChipCloudConfig;
import io.noties.markwon.Markwon;
import io.noties.markwon.movement.MovementMethodPlugin;
import io.realm.Realm;

public class AdapterCourses extends RecyclerView.Adapter<AdapterCourses.ViewHoldercourse> {
    private RowCourseBinding rowCourseBinding;
    private Context context;
    private List<RealmMyCourse> courseList;
    private List<RealmMyCourse> selectedItems;
    private OnCourseItemSelected listener;
    private OnHomeItemClickListener homeItemClickListener;
    private HashMap<String, JsonObject> map;
    private HashMap<String, JsonObject> progressMap;
    private OnRatingChangeListener ratingChangeListener;
    private Realm mRealm;
    private ChipCloudConfig config;
    private Markwon markwon;
    private boolean isAscending = true;
    private boolean isTitleAscending = true;
    private boolean areAllSelected = true;

    public AdapterCourses(Context context, List<RealmMyCourse> courseList, HashMap<String, JsonObject> map) {
        this.map = map;
        this.context = context;
        this.courseList = courseList;
        markwon = Markwon.builder(context)
                .usePlugin(MovementMethodPlugin.none())
                .build();
        this.selectedItems = new ArrayList<>();
        if (context instanceof OnHomeItemClickListener) {
            homeItemClickListener = (OnHomeItemClickListener) context;
        }
        config = Utilities.getCloudConfig().selectMode(ChipCloud.SelectMode.single);
    }

    public static void showRating(JsonObject object, TextView average, TextView ratingCount, AppCompatRatingBar ratingBar) {
        average.setText(String.format("%.2f", object.get("averageRating").getAsFloat()));
        ratingCount.setText(object.get("total").getAsInt() + " total");
        if (object.has("ratingByUser")) ratingBar.setRating(object.get("ratingByUser").getAsInt());
        else ratingBar.setRating(0);
    }

    public void setmRealm(Realm mRealm) {
        this.mRealm = mRealm;
    }

    public void setRatingChangeListener(OnRatingChangeListener ratingChangeListener) {
        this.ratingChangeListener = ratingChangeListener;
    }

    public List<RealmMyCourse> getCourseList() {
        return courseList;
    }

    public void setCourseList(List<RealmMyCourse> courseList) {
        this.courseList = courseList;
        sortCourseList();
        sortCourseListByTitle();
        notifyDataSetChanged();
    }

    private void sortCourseListByTitle() {
        Collections.sort(courseList, (course1, course2) -> {
            if (isTitleAscending) {
                return course1.getCourseTitle().compareToIgnoreCase(course2.getCourseTitle());
            } else {
                return course2.getCourseTitle().compareToIgnoreCase(course1.getCourseTitle());
            }
        });
    }

    private void sortCourseList() {
        Collections.sort(courseList, new Comparator<RealmMyCourse>() {
            @Override
            public int compare(RealmMyCourse course1, RealmMyCourse course2) {
                if (isAscending) {
                    return course1.getCreatedDate().compareTo(course2.getCreatedDate());
                } else {
                    return course2.getCreatedDate().compareTo(course1.getCreatedDate());
                }
            }
        });
    }

    public void toggleTitleSortOrder() {
        isTitleAscending = !isTitleAscending;
        sortCourseListByTitle();
        notifyDataSetChanged();
    }

    public void toggleSortOrder() {
        isAscending = !isAscending;
        sortCourseList();
        notifyDataSetChanged();
    }

    public void setProgressMap(HashMap<String, JsonObject> progressMap) {
        this.progressMap = progressMap;
    }

    public void setListener(OnCourseItemSelected listener) {
        this.listener = listener;
    }

    @NonNull
    @Override
    public ViewHoldercourse onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
//        View v = LayoutInflater.from(context).inflate(R.layout.row_course, parent, false);
//        return new ViewHoldercourse(v);
        rowCourseBinding = RowCourseBinding.inflate(LayoutInflater.from(context), parent, false);
        return new ViewHoldercourse(rowCourseBinding);
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHoldercourse holder, final int position) {
        ViewHoldercourse viewHolder = (ViewHoldercourse) holder;
        viewHolder.bind(position);
        rowCourseBinding.title.setText(courseList.get(position).getCourseTitle());
        rowCourseBinding.description.setText(courseList.get(position).getDescription());
        markwon.setMarkdown(rowCourseBinding.description, courseList.get(position).getDescription());

        rowCourseBinding.gradLevel.setText(context.getString(R.string.grade_level_colon) + courseList.get(position).getGradeLevel());
        rowCourseBinding.subjectLevel.setText(context.getString(R.string.subject_level_colon) + courseList.get(position).getSubjectLevel());
        rowCourseBinding.checkbox.setChecked(selectedItems.contains(courseList.get(position)));
        rowCourseBinding.courseProgress.setMax(courseList.get(position).getnumberOfSteps());
        displayTagCloud(rowCourseBinding.flexboxDrawable, position);
        try {
            rowCourseBinding.tvDate.setText(TimeUtils.formatDate(Long.parseLong(courseList.get(position).getCreatedDate().trim()), "MMM dd, yyyy"));
        } catch (Exception e) {

        }
        rowCourseBinding.ratingBar.setOnTouchListener((v1, event) -> {
            if (event.getAction() == MotionEvent.ACTION_UP)
                homeItemClickListener.showRatingDialog("course", courseList.get(position).getCourseId(), courseList.get(position).getCourseTitle(), ratingChangeListener);
            return true;
        });

        rowCourseBinding.checkbox.setOnClickListener((view) -> {
            Utilities.handleCheck(((CheckBox) view).isChecked(), position, (ArrayList) selectedItems, courseList);
            if (listener != null) listener.onSelectedListChange(selectedItems);
            notifyDataSetChanged();
        });
        showProgressAndRating(position, holder);
    }

    public boolean areAllSelected(){
        if (selectedItems.size() != courseList.size()) {
            areAllSelected = false;
        }
        return areAllSelected;
    }

    public void selectAllItems(boolean selectAll) {
        if (selectAll) {
            selectedItems.clear();
            selectedItems.addAll(courseList);
        } else {
            selectedItems.clear();
        }

        notifyDataSetChanged();

        if (listener != null) {
            listener.onSelectedListChange(selectedItems);
        }
    }


    private void displayTagCloud(FlexboxLayout flexboxDrawable, int position) {
        flexboxDrawable.removeAllViews();
        final ChipCloud chipCloud = new ChipCloud(context, flexboxDrawable, config);
        List<RealmTag> tags = mRealm.where(RealmTag.class).equalTo("db", "courses").equalTo("linkId", courseList.get(position).getId()).findAll();
        showTags(tags, chipCloud);
    }

    private void showTags(List<RealmTag> tags, ChipCloud chipCloud) {
        for (RealmTag tag : tags) {
            RealmTag parent = mRealm.where(RealmTag.class).equalTo("id", tag.getTagId()).findFirst();
            showChip(chipCloud, parent);
        }
    }

    private void showChip(ChipCloud chipCloud, RealmTag parent) {
        chipCloud.addChip(((parent != null) ? parent.getName() : ""));
        chipCloud.setListener((i, b, b1) -> {
            if (b1 && listener != null) {
                listener.onTagClicked(parent);
            }
        });
    }

    private void showProgressAndRating(int position, ViewHoldercourse holder) {
        showProgress(position, holder);
        if (map.containsKey(courseList.get(position).getCourseId())) {
            JsonObject object = map.get(courseList.get(position).getCourseId());
            showRating(object, rowCourseBinding.average, rowCourseBinding.timesRated, rowCourseBinding.ratingBar);
        } else {
            rowCourseBinding.ratingBar.setRating(0);
        }
    }

    private void showProgress(int position, ViewHoldercourse holder) {
        if (progressMap.containsKey(courseList.get(position).getCourseId())) {
            JsonObject ob = progressMap.get(courseList.get(position).getCourseId());
            rowCourseBinding.courseProgress.setMax(JsonUtils.getInt("max", ob));
            rowCourseBinding.courseProgress.setProgress(JsonUtils.getInt("current", ob));
            if (JsonUtils.getInt("current", ob) < JsonUtils.getInt("max", ob))
                rowCourseBinding.courseProgress.setSecondaryProgress(JsonUtils.getInt("current", ob) + 1);
            rowCourseBinding.courseProgress.setVisibility(View.VISIBLE);
        } else {
            rowCourseBinding.courseProgress.setVisibility(View.GONE);
        }
    }

    private void openCourse(RealmMyCourse realm_myCourses, int i) {
        if (homeItemClickListener != null) {
            Fragment f = new TakeCourseFragment();
            Bundle b = new Bundle();
            b.putString("id", realm_myCourses.getCourseId());
            b.putInt("position", i);
            f.setArguments(b);
            homeItemClickListener.openCallFragment(f);
        }
    }

    @Override
    public int getItemCount() {
        return courseList.size();
    }

    class ViewHoldercourse extends RecyclerView.ViewHolder {
//        TextView title, desc, grad_level, subject_level, tvDate, ratingCount, average;
//        CheckBox checkBox;
//        AppCompatRatingBar ratingBar;
//        SeekBar progressBar;
//        LinearLayout llRating;
//        FlexboxLayout flexboxLayout;
        RowCourseBinding rowCourseBinding;

        private int adapterPosition;

        public ViewHoldercourse(RowCourseBinding rowCourseBinding) {
            super(rowCourseBinding.getRoot());
            this.rowCourseBinding = rowCourseBinding;
            itemView.setOnClickListener(v -> {
                if (adapterPosition != RecyclerView.NO_POSITION) {
                    openCourse(courseList.get(adapterPosition), 0);
                }
            });

            if (Build.VERSION.SDK_INT <= Build.VERSION_CODES.LOLLIPOP) {
                rowCourseBinding.courseProgress.setScaleY(0.3f);
            }

            rowCourseBinding.courseProgress.setOnSeekBarChangeListener(new SeekBar.OnSeekBarChangeListener() {
                @Override
                public void onProgressChanged(SeekBar seekBar, int i, boolean b) {
                    if (progressMap.containsKey(courseList.get(getAdapterPosition()).getCourseId())) {
                        JsonObject ob = progressMap.get(courseList.get(getAdapterPosition()).getCourseId());
                        int current = JsonUtils.getInt("current", ob);
                        if (b && i <= current + 1) {
                            openCourse(courseList.get(getAdapterPosition()), seekBar.getProgress());
                        }
                    }
                }

                @Override
                public void onStartTrackingTouch(SeekBar seekBar) {

                }

                @Override
                public void onStopTrackingTouch(SeekBar seekBar) {

                }
            });
        }

        public void bind(int position) {
            adapterPosition = position; // Store the adapter position
        }
    }
}

File: app/src/main/java/org/ole/planet/myplanet/ui/courses/AdapterCourses.kt
package org.ole.planet.myplanet.ui.courses

import android.content.Context
import android.os.Bundle
import android.view.LayoutInflater
import android.view.MotionEvent
import android.view.View
import android.view.ViewGroup
import android.widget.CheckBox
import android.widget.SeekBar
import android.widget.SeekBar.OnSeekBarChangeListener
import android.widget.TextView
import androidx.fragment.app.Fragment
import androidx.lifecycle.findViewTreeLifecycleOwner
import androidx.lifecycle.lifecycleScope
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView
import com.google.android.flexbox.FlexboxLayout
import com.google.gson.JsonObject
import fisk.chipcloud.ChipCloud
import fisk.chipcloud.ChipCloudConfig
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import org.ole.planet.myplanet.R
import org.ole.planet.myplanet.callback.OnCourseItemSelected
import org.ole.planet.myplanet.callback.OnHomeItemClickListener
import org.ole.planet.myplanet.callback.OnRatingChangeListener
import org.ole.planet.myplanet.databinding.RowCourseBinding
import org.ole.planet.myplanet.model.RealmTag
import org.ole.planet.myplanet.model.RealmUserModel
import org.ole.planet.myplanet.model.dto.CourseItem
import org.ole.planet.myplanet.repository.TagRepository
import org.ole.planet.myplanet.utilities.CourseRatingUtils
import org.ole.planet.myplanet.utilities.JsonUtils.getInt
import org.ole.planet.myplanet.utilities.Markdown.prependBaseUrlToImages
import org.ole.planet.myplanet.utilities.Markdown.setMarkdownText
import org.ole.planet.myplanet.utilities.SelectionUtils
import org.ole.planet.myplanet.utilities.TimeUtils.formatDate
import org.ole.planet.myplanet.utilities.Utilities

class AdapterCourses(
    private val context: Context,
    private var userModel: RealmUserModel?,
    private val tagRepository: TagRepository
) : ListAdapter<CourseItem, RecyclerView.ViewHolder>(CourseDiffCallback) {
    private val selectedItems: MutableList<CourseItem?> = ArrayList()
    private var listener: OnCourseItemSelected? = null
    private var homeItemClickListener: OnHomeItemClickListener? = null
    private var ratingChangeListener: OnRatingChangeListener? = null
    private val config: ChipCloudConfig
    private var areAllSelected = false
    private val tagCache: MutableMap<String, List<RealmTag>> = mutableMapOf()
    private val tagRequestsInProgress: MutableSet<String> = mutableSetOf()

    companion object {
        private const val TAG_PAYLOAD = "payload_tags"

        private val CourseDiffCallback = object : DiffUtil.ItemCallback<CourseItem>() {
            override fun areItemsTheSame(oldItem: CourseItem, newItem: CourseItem): Boolean {
                return oldItem.id == newItem.id
            }

            override fun areContentsTheSame(oldItem: CourseItem, newItem: CourseItem): Boolean {
                return oldItem == newItem
            }
        }
    }

    init {
        if (context is OnHomeItemClickListener) {
            homeItemClickListener = context
        }
        config = Utilities.getCloudConfig().selectMode(ChipCloud.SelectMode.single)
    }

    fun setRatingChangeListener(ratingChangeListener: OnRatingChangeListener?) {
        this.ratingChangeListener = ratingChangeListener
    }

    fun setListener(listener: OnCourseItemSelected?) {
        this.listener = listener
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        val binding = RowCourseBinding.inflate(LayoutInflater.from(parent.context), parent, false)
        return ViewHoldercourse(binding)
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        if (holder !is ViewHoldercourse) return

        holder.bind(position)
        val course = getItem(position) ?: return

        updateVisibilityForMyCourse(holder, course)
        holder.rowCourseBinding.title.text = course.courseTitle
        configureDescription(holder, course, position)
        configureDateViews(holder, course)
        setTextViewContent(
            holder.rowCourseBinding.gradLevel,
            course.gradeLevel,
            holder.rowCourseBinding.gradLevel,
            context.getString(R.string.grade_level_colon)
        )
        setTextViewContent(
            holder.rowCourseBinding.subjectLevel,
            course.subjectLevel,
            holder.rowCourseBinding.subjectLevel,
            context.getString(R.string.subject_level_colon)
        )
        holder.rowCourseBinding.courseProgress.max = course.numberOfSteps
        displayTagCloud(holder, position)

        val isGuest = userModel?.isGuest() ?: true
        if (!isGuest) setupRatingBar(holder, course)
        setupCheckbox(holder, course, position, isGuest)

        updateRatingViews(holder, position)
        updateProgressViews(holder, position)

        holder.rowCourseBinding.root.setOnClickListener {
            val newPosition = holder.bindingAdapterPosition
            if (newPosition != RecyclerView.NO_POSITION) {
                openCourse(getItem(newPosition), 0)
            }
        }
    }

    private fun updateVisibilityForMyCourse(holder: ViewHoldercourse, course: CourseItem) {
        if (course.isMyCourse) {
            holder.rowCourseBinding.isMyCourse.visibility = View.VISIBLE
            holder.rowCourseBinding.checkbox.visibility = View.GONE
        } else {
            holder.rowCourseBinding.isMyCourse.visibility = View.GONE
            holder.rowCourseBinding.checkbox.visibility = View.VISIBLE
        }
    }

    private fun configureDescription(holder: ViewHoldercourse, course: CourseItem, position: Int) {
        holder.rowCourseBinding.description.apply {
            text = course.description
            val markdownContentWithLocalPaths = prependBaseUrlToImages(
                course.description,
                "file://${context.getExternalFilesDir(null)}/ole/",
                150,
                100
            )
            setMarkdownText(this, markdownContentWithLocalPaths)

            setOnClickListener {
                homeItemClickListener?.openCallFragment(TakeCourseFragment().apply {
                    arguments = Bundle().apply {
                        putString("id", course.courseId)
                        putInt("position", position)
                    }
                })
            }
        }
    }

    private fun configureDateViews(holder: ViewHoldercourse, course: CourseItem) {
        if (course.gradeLevel.isNullOrEmpty() && course.subjectLevel.isNullOrEmpty()) {
            holder.rowCourseBinding.holder.visibility = View.VISIBLE
            holder.rowCourseBinding.tvDate2.visibility = View.VISIBLE
            holder.rowCourseBinding.tvDate.visibility = View.GONE
            try {
                holder.rowCourseBinding.tvDate2.text = formatDate(course.createdDate ?: 0, "MMM dd, yyyy")
            } catch (e: Exception) {
                throw RuntimeException(e)
            }
        } else {
            holder.rowCourseBinding.tvDate.visibility = View.VISIBLE
            holder.rowCourseBinding.tvDate2.visibility = View.GONE
            holder.rowCourseBinding.holder.visibility = View.GONE
            try {
                holder.rowCourseBinding.tvDate.text = formatDate(course.createdDate ?: 0, "MMM dd, yyyy")
            } catch (e: Exception) {
                throw RuntimeException(e)
            }
        }
    }

    private fun setupRatingBar(holder: ViewHoldercourse, course: CourseItem) {
        holder.rowCourseBinding.ratingBar.setOnTouchListener { _: View?, event: MotionEvent ->
            if (event.action == MotionEvent.ACTION_UP) homeItemClickListener?.showRatingDialog(
                "course",
                course.courseId,
                course.courseTitle,
                ratingChangeListener
            )
            true
        }
    }

    private fun setupCheckbox(holder: ViewHoldercourse, course: CourseItem, position: Int, isGuest: Boolean) {
        if (!isGuest) {
            if (course.isMyCourse) {
                holder.rowCourseBinding.checkbox.visibility = View.GONE
            } else {
                holder.rowCourseBinding.checkbox.visibility = View.VISIBLE
                holder.rowCourseBinding.checkbox.isChecked = selectedItems.contains(course)
                holder.rowCourseBinding.checkbox.setOnClickListener { view: View ->
                    holder.rowCourseBinding.checkbox.contentDescription =
                        context.getString(R.string.select_res_course, course.courseTitle)
                    SelectionUtils.handleCheck((view as CheckBox).isChecked, position, selectedItems, currentList)
                    listener?.onSelectedListChange(selectedItems)
                }
            }
        } else {
            holder.rowCourseBinding.checkbox.visibility = View.GONE
        }
    }

    private fun setTextViewContent(textView: TextView?, content: String?, layout: View?, prefix: String) {
        if (content.isNullOrEmpty()) {
            layout?.visibility = View.GONE
        } else {
            layout?.visibility = View.VISIBLE
            textView?.text = context.getString(R.string.prefix_content, prefix, content)
        }
    }

    fun areAllSelected(): Boolean {
        val selectableCourses = currentList.filterNotNull().filter { !it.isMyCourse }
        areAllSelected = selectedItems.size == selectableCourses.size && selectableCourses.isNotEmpty()
        return areAllSelected
    }

    fun selectAllItems(selectAll: Boolean) {
        selectedItems.clear()

        if (selectAll) {
            val selectableCourses = currentList.filterNotNull().filter { !it.isMyCourse }
            selectedItems.addAll(selectableCourses)
        }

        val updatedPositions = mutableListOf<Int>()
        currentList.forEachIndexed { index, course ->
            if (course != null && !course.isMyCourse) {
                updatedPositions.add(index)
            }
        }

        updatedPositions.forEach { position ->
            notifyItemChanged(position)
        }

        listener?.onSelectedListChange(selectedItems)
    }

    override fun onBindViewHolder(
        holder: RecyclerView.ViewHolder,
        position: Int,
        payloads: MutableList<Any>
    ) {
        if (holder !is ViewHoldercourse) {
            super.onBindViewHolder(holder, position, payloads)
            return
        }

        val hasTagPayload = payloads.any { it == TAG_PAYLOAD }
        if (payloads.isNotEmpty() && hasTagPayload) {
            val courseId = getItem(position)?.id ?: return
            val tags = tagCache[courseId].orEmpty()
            renderTagCloud(holder.rowCourseBinding.flexboxDrawable, tags)
        } else {
            super.onBindViewHolder(holder, position, payloads)
        }
    }


    private fun displayTagCloud(holder: ViewHoldercourse, position: Int) {
        val flexboxDrawable = holder.rowCourseBinding.flexboxDrawable
        val courseId = getItem(position)?.id
        if (courseId == null) {
            flexboxDrawable.removeAllViews()
            return
        }

        val cachedTags = tagCache[courseId]
        if (cachedTags != null) {
            renderTagCloud(flexboxDrawable, cachedTags)
            return
        }

        flexboxDrawable.removeAllViews()

        if (!tagRequestsInProgress.add(courseId)) {
            return
        }

        holder.itemView.findViewTreeLifecycleOwner()?.lifecycleScope?.launch {
            try {
                val tags = tagRepository.getTagsForCourse(courseId)
                tagCache[courseId] = tags
                val adapterPosition = holder.bindingAdapterPosition
                if (adapterPosition != RecyclerView.NO_POSITION) {
                    withContext(kotlinx.coroutines.Dispatchers.Main) {
                        notifyItemChanged(adapterPosition, TAG_PAYLOAD)
                    }
                }
            } finally {
                tagRequestsInProgress.remove(courseId)
            }
        }
    }

    private fun renderTagCloud(flexboxDrawable: FlexboxLayout, tags: List<RealmTag>) {
        flexboxDrawable.removeAllViews()
        if (tags.isEmpty()) {
            return
        }
        val chipCloud = ChipCloud(context, flexboxDrawable, config)
        tags.forEach { tag ->
            chipCloud.addChip(tag.name ?: "")
        }
        chipCloud.setListener { index: Int, _: Boolean, isSelected: Boolean ->
            if (isSelected) {
                tags.getOrNull(index)?.let { selectedTag ->
                    listener?.onTagClicked(selectedTag)
                }
            }
        }
    }

    private fun updateRatingViews(holder: ViewHoldercourse, position: Int) {
        val course = getItem(position) ?: return
        val averageRating = course.averageRating
        val totalRatings = course.totalRatings
        if (averageRating != null && totalRatings != null) {
            holder.rowCourseBinding.ratingBar.rating = averageRating
            holder.rowCourseBinding.rating.text = context.getString(R.string.rating_format, averageRating)
            holder.rowCourseBinding.timesRated.text = context.getString(R.string.rating_count_format, totalRatings)
        } else {
            holder.rowCourseBinding.ratingBar.rating = 0f
            holder.rowCourseBinding.rating.text = context.getString(R.string.zero_point_zero)
            holder.rowCourseBinding.timesRated.text = context.getString(R.string.rating_count_format, 0)
        }
    }

    private fun updateProgressViews(holder: ViewHoldercourse, position: Int) {
        val course = getItem(position) ?: return
        val currentProgress = course.currentProgress
        val maxProgress = course.maxProgress
        if (currentProgress != null && maxProgress != null) {
            holder.rowCourseBinding.courseProgress.max = maxProgress
            holder.rowCourseBinding.courseProgress.progress = currentProgress
            if (currentProgress < holder.rowCourseBinding.courseProgress.max) {
                holder.rowCourseBinding.courseProgress.secondaryProgress = currentProgress + 1
            }
            holder.rowCourseBinding.courseProgress.visibility = View.VISIBLE
        } else {
            holder.rowCourseBinding.courseProgress.visibility = View.GONE
        }
    }

    private fun openCourse(realmMyCourses: CourseItem?, step: Int) {
        if (homeItemClickListener != null) {
            val f: Fragment = TakeCourseFragment()
            val b = Bundle()
            b.putString("id", realmMyCourses?.courseId)
            b.putInt("position", step)
            f.arguments = b
            homeItemClickListener?.openCallFragment(f)
        }
    }

    internal inner class ViewHoldercourse(val rowCourseBinding: RowCourseBinding) :
        RecyclerView.ViewHolder(rowCourseBinding.root) {
        private var adapterPosition = 0

        init {
            itemView.setOnClickListener {
                if (adapterPosition != RecyclerView.NO_POSITION) {
                    openCourse(getItem(adapterPosition), 0)
                }
            }
            rowCourseBinding.courseProgress.scaleY = 0.3f
            rowCourseBinding.courseProgress.setOnSeekBarChangeListener(object : OnSeekBarChangeListener {
                override fun onProgressChanged(seekBar: SeekBar, i: Int, b: Boolean) {
                    val position = bindingAdapterPosition
                    if (position != RecyclerView.NO_POSITION) {
                        val course = getItem(position)
                        val current = course.currentProgress
                        if (b && current != null && i <= current + 1) {
                            openCourse(getItem(bindingAdapterPosition), seekBar.progress)
                        }
                    }
                }

                override fun onStartTrackingTouch(seekBar: SeekBar) {}
                override fun onStopTrackingTouch(seekBar: SeekBar) {}
            })
        }

        fun bind(position: Int) {
            adapterPosition = position
        }
    }
}
